framework/src/materials/GenericConstantFunctorMaterial.C

```diff
@@ -0,0 +1,55 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "GenericConstantFunctorMaterial.h"
+
+registerMooseObject("MooseApp", GenericConstantFunctorMaterial);
+registerMooseObject("MooseApp", ADGenericConstantFunctorMaterial);
+
+template <bool is_ad>
+InputParameters
+GenericConstantFunctorMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription(
+      "Declares material properties based on names and values prescribed by input parameters.");
+  params.addRequiredParam<std::vector<std::string>>(
+      "prop_names", "The names of the properties this material will have");
+  params.addRequiredParam<std::vector<Real>>("prop_values",
+                                             "The values associated with the named properties");
+  return params;
+}
+
+template <bool is_ad>
+GenericConstantFunctorMaterialTempl<is_ad>::GenericConstantFunctorMaterialTempl(
+    const InputParameters & parameters)
+  : Material(parameters),
+    _prop_names(getParam<std::vector<std::string>>("prop_names")),
+    _prop_values(getParam<std::vector<Real>>("prop_values"))
+{
+  unsigned int num_names = _prop_names.size();
+  unsigned int num_values = _prop_values.size();
+
+  if (num_names != num_values)
+    mooseError("Number of prop_names must match the number of prop_values for a "
+               "GenericConstantFunctorMaterial!");
+
+  _num_props = num_names;
+
+  for (const auto i : make_range(_num_props))
+  {
+    auto & prop = declareFunctorProperty<GenericReal<is_ad>>(_prop_names[i]);
+    prop.setFunction(_mesh, blockIDs(), [this, i](auto & /*geom_quantity*/) -> GenericReal<is_ad> {
+      return _prop_values[i];
+    });
```

Comment:
What is the failure mode if someone calls setFunction with a lambda that doesn't work for all our expected lambda args -  a bunch of nasty compiler errors right?  If a user wants to create a material property that only supports a subset of mesh locations, do we expect them to write their own FunctorInterface subclass implementation?  I think it would be nice for us to offer a better path for this.  Or maybe I'm misunderstanding how all this works together.